--local option for 'connect' to connect through Local Relay

rafasofizada opened the pull request.

rafasofizada committed:
Local Relay integration; --local connect option to spawn Local Relay instance

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/commands/connect.ts:
```diff
@@ -71,14 +79,104 @@ export const connectAction = async (
+  // Dangers of POSIX fork() (https://www.evanjones.ca/fork-is-dangerous.html) don't apply.
+  child_process.fork(relayJsPath, {
+    detached: true,
+    // We don't want to read parent's stdin from child process, but we want to share the same stdout/stderr.
```
What happens once the shell is closed and that stdin/stderr no longer exists? Piping stdout to a log file may be more appropriate.

Hexxeh commented:
Note that the commands previously contain very little actual logic, instead calling out to a different file. Typically the commands themselves contain only argument validation/sanity checks, and call out to a file under models for their actual business logic.

Doing this will make your life simpler when you write tests to go with this PR.

rafasofizada committed:
apiFetch – configure API URLs and auth requirement

Need help with "ASK"s in baseAPI.test.ts

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor baseApi test refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Refactor DeveloperRelay into class and allow custom API URL, auth flow

1. DeveloperRelay is now a class instance instead of simply a module.
2. Point 1 enables setting custom API URLs and skipping authentification when needed.
3. developerRelay tests test the module functionality with different configurations
4. HostConnections methods accepts custom DeveloperRelay instances.

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Remove unnecessary developerRelay mock

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Problems testing LocalRelay

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
localRelay model, tests for relayInfo.readRelayInfo, util

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Refactoring and comment fixes

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo+test – minor refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
readJsonFile typing

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Stray empty line

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
readRelayInfo better error handling

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
HostConnections.connect – minor refactor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
HostConnections test – minor refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Use p-wait-for for pollRelayInfo

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Refactor readJsonFile typing

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor refactor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
40ad69e – connect.test.ts fix

Solution: mock `developerRelay.connect()`

Problem: during connect tests, `developerRelay.connect()` fails (which is expected, there's no real WS URL to connect to). Before 40ad69e this didn't matter, as the function was called from within `hostConnections.connect()`, and all that was tested was whether `hostConnections.connect()` was called with correct arguments – whatever errors thrown inside didn't matter. After 40ad69e, the two function calls got separated, so `developerRelay.connect()` throwing prevented `hostConnections.connect()` from being called.

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo test fix

Addition to f57bdf18648725cee77d859356867a329ffada5f

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor import rename

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada commented on packages/sdk-cli/src/models/localRelay/util.ts:
```diff
@@ -0,0 +1,10 @@
+import { readFile } from 'fs/promises';
+
+export async function readJsonFile<T = unknown>(path: string): Promise<T> {
```
https://github.com/Fitbit/developer-bridge/pull/176/commits/3302789fd39914620162dd94636f636fb5a79c65

rafasofizada commented on packages/sdk-cli/src/models/localRelay/relayInfo.ts:
```diff
@@ -0,0 +1,93 @@
+
+  if (!entryPoint || entryPoint === '') {
+    throw new Error(
+      `No 'main' path specified in ${RELAY_PKG_NAME}'s package.json (${fullPath})`,
```
https://github.com/Fitbit/developer-bridge/pull/176/commits/3302789fd39914620162dd94636f636fb5a79c65

rafasofizada commented on packages/sdk-cli/src/models/localRelay/relayInfo.ts:
```diff
@@ -0,0 +1,93 @@
+  return false;
+}
+
+export async function pollRelayInfo(
```
https://github.com/Fitbit/developer-bridge/pull/176/commits/f57bdf18648725cee77d859356867a329ffada5f
https://github.com/Fitbit/developer-bridge/pull/176/commits/a463f1d61a6b415d4fedf0d1720c83e548e49038

rafasofizada committed:
Pipe child process' output/errors to a log file

There are a couple of issues marked with [CONFUSING] and [TODO}, needs some experimentation with Node child_process and fs api to figure this out

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -0,0 +1,30 @@
+  // Dangers of POSIX fork() (https://www.evanjones.ca/fork-is-dangerous.html) don't apply.
+  child_process.fork(relayJsPath, {
+    detached: true,
+    // We don't want to read parent's stdin from child process, but we want to share the same stdout/stderr.
```
https://github.com/Fitbit/developer-bridge/pull/176/commits/07a984fbd4e2e4dc201926248342bd77d637aa39

There are a couple of issues marked with [CONFUSING] and [TODO}, needs some experimentation with Node child_process and fs api to figure this out

rafasofizada replied:
> what happens if the parent process dies, does the child live on, even though these pipes close? Have you tested?

Really confusing. Practically speaking, as long as `detached: true` is set, child process will live regardless of the parent (tested this). However, the documentation seems to disagree: https://nodejs.org/api/child_process.html#optionsdetached

> "When using the detached option to start a long-running process, the process will not stay running in the background after the parent exits unless it is provided with a stdio configuration that is not connected to the parent. If the parent's stdio is inherited, the child will remain attached to the controlling terminal."

However, I can do:
 ```
child_process.fork('./test.js', { stdio: ['ignore', 'inherit', 'inherit', 'ipc'], detached: true });
                                                     ^^^^^^^^^^^^^^^^^^^^          ^^^^^^^^
```

And everything will work as wanted, despite the `inherit` stdio option conflicting with `detached: true` requirement: "unless it is provided with a **stdio configuration** that is **not connected** to the parent"

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/07a984fbd4e2e4dc201926248342bd77d637aa39
https://github.com/Fitbit/developer-bridge/pull/176/commits/5c44f5f50f13fa13437b54ea5c5b76c31239d3fe

rafasofizada replied:
Not sure how to pipe child process' output both to parent.process.stdout and the log file. Any ideas? Passthrough stream?

Hexxeh replied:
For the reasons I noted in another comment, I wouldn't pipe it to stdout (ie, it would interleave cli-sdk output confusingly potentially).

Hexxeh replied:
Seems like there's also some concerns to deal with re https://nodejs.org/api/child_process.html#subprocessunref too

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/b22d14be4d2f537d208b17320e350a2b79d26786

rafasofizada replied:
The rabbithole actually goes deeper, it is possible to use `fork()` but separately `unref()` the mandatory ipc channel. The docs say it is actually done by default for `spawn()` (which has an IPC channel only if you specify it in `options.stdio`), https://nodejs.org/api/child_process.html#optionsstdio:

> It is worth noting that when an IPC channel is established between the parent and child processes, and the child
> is a Node.js process, the child is launched with the IPC channel unreferenced (using unref()) until the child 
> registers an event handler for the 'disconnect' event or the 'message' event. This allows the child to exit normally
> without the process being held open by the open IPC channel.

No idea why this isn't the case with `fork()`. Tried digging in the source code for fun but to no avail (for now:d).

Hexxeh replied:
Could you maybe just use `spawn()` then if it has the desired semantics?

Hexxeh replied:
I see you have, great :)

rafasofizada committed:
Better comments for forking Relay child process

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/babel.config.js:
```diff
@@ -0,0 +1 @@
+module.exports = { presets: ['@babel/preset-env'] };
```
Is this necessary? Would think it'd use the file in the root. I don't see it in my example here: https://github.com/Fitbit/developer-bridge/commit/22f93b1e8f61cdcd13c05c7ac9a5d40c32dbb519

Hexxeh commented on packages/sdk-cli/src/models/DeveloperRelay.ts:
```diff
@@ -0,0 +1,96 @@
+
+export default class DeveloperRelay {
+  constructor(
+    public readonly apiUrl: string = environment().config.apiUrl,
```
These can be private

Hexxeh commented on packages/sdk-cli/src/models/HostConnections.ts:
```diff
@@ -64,11 +61,11 @@ class HostConnections {
   companionHost?: HostConnection;
 
-  async connect(hostType: HostType, hostID: string) {
+  async connect(hostType: HostType, hostWs: stream.Duplex) {
```
nit: technically you don't care this is a WebSocket so `hostStream` might be more appropriate of a name

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -0,0 +1,60 @@
+
+  // Official docs use fs.openSync, which returns a file descriptor integer.
+  // In our case, using a Writable Stream is more intuitive.
+  const out = createWriteStream(RELAY_LOG_FILE_PATH, { flags: 'a' });
```
Can you not just open once and pass the same reference to both?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.ts:
```diff
@@ -0,0 +1,67 @@
+export async function pollRelayInfo(
+  timeout = 15000,
+  interval = 300,
+  readRelayInfoFn = readRelayInfo,
```
Why does this need to be an argument? The extra indirection just makes it a little harder to follow.

rafasofizada replied:
In `pollRelayInfo` tests I sometimes need to mock `readRelayInfo`. And I don't really want to separate `readRelayInfo.ts` into 2 separate files

Hexxeh replied:
Just mock the file it's reading underneath?

rafasofizada replied:
I test for the number of times `readRelayInfoFn()` has been called, too

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/27f8d0133fbe6519f135b8f278e0f089a5aa60b3

Hexxeh commented on packages/sdk-cli/src/api/baseAPI.test.ts:
```diff
@@ -23,6 +24,30 @@ describe('apiFetch()', () => {
+    const fakeAPIDomain = 'https://fake-dev-relay-test-endpoint.fitbit.com';
+
+    mockAuthToken(null);
+    // apiFetch would throw if it didn't skip auth, because authToken is falsy
```
you can actually check this, rather than awaiting it, use `await expect(response).resolves.toBeDefined()`

rafasofizada replied:
if on the above line I assigned an `awaited` value to response, the response is no longer a Promise, right? So I can just do `expect(response).toBeDefined()`

Hexxeh replied:
You can, but I think Jest gives a better error if you let it handle the resolution/rejection if it doesn't resolve as expected.

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/26ee1d9e19334aedfd889137a6271b563b290a65 ?

Hexxeh replied:
Yep, looks fine.

Hexxeh commented on packages/sdk-cli/src/commands/connect.test.ts:
```diff
@@ -3,41 +3,43 @@ import events from 'events';
 import commandTestHarness from '../testUtils/commandTestHarness';
-import * as developerRelay from '../api/developerRelay';
 import HostConnections, { HostType } from '../models/HostConnections';
+import { Duplex } from 'stream';
```
nit: move non project import to top

Hexxeh commented on packages/sdk-cli/src/commands/connect.test.ts:
```diff
@@ -147,6 +160,18 @@ describe.each<[DeviceType, HostType]>([
+    const mockHost = mockRelayHosts[deviceType][0];
+    mockRelayHostsResponse[deviceType]([mockHost]);
+
+    const port = 1;
```
nit: inline as `port: 1`?

Hexxeh replied:
Noticed you'd resolved this but you've done the opposite of what I suggested?

Hexxeh commented on packages/sdk-cli/src/models/HostConnections.test.ts:
```diff
@@ -23,12 +24,13 @@ function mockSentinel(spy: jest.SpyInstance) {
-function doConnect(type: HostType) {
-  return hostConnections.connect(type, mockHostID);
+async function doConnect(type: HostType) {
+  const ws = await relayInstance.connect(mockHostID);
```
Can this just use a mock or a `stream.Duplex`?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -0,0 +1,196 @@
+import { TimeoutError } from 'p-timeout';
+import { join } from 'path';
+import { cwd } from 'process';
+import { RELAY_PKG_NAME } from './const';
```
nit: split system/project imports with a blank line

Hexxeh commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -0,0 +1,89 @@
+  };
+});
+
+describe('isInt', () => {
```
nit: seems like a good spot to use `.each`

Hexxeh commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -0,0 +1,89 @@
+  });
+});
+
+describe('readJsonFile', () => {
```
Why use a mock FS for some cases and a real FS for others?

rafasofizada replied:
The test cases are actually identical, I just lost track of the case names.

rafasofizada replied:
I tried to `each` -ify them, but the test cases have subtle differences which make this not worth the effort

Hexxeh replied:
Okay, but why test a mock and real FS separately?

rafasofizada replied:
Honestly speaking, I don't know. I just feel much safer with having tests using real FS; like, if I were to choose, I'd leave the "real" tests and remove the mock-fs ones. I think of real-fs cases as a test of whether the assumptions I put into the mock-fs tests are correct

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/d08907b3bf6baff82219004f41e50833185d3a61

rafasofizada committed:
Delete unnecessary babel.config.js

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Local Relay child_process call refactoring (spawn instead of fork)

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada commented on packages/sdk-cli/src/commands/connect.test.ts:
```diff
@@ -56,22 +58,29 @@ let mockWS: events.EventEmitter;
   hostConnectSpy = jest.spyOn(hostConnections, 'connect');
+  relayConnectSpy = jest.spyOn(DeveloperRelay.prototype, 'connect');
+
+  // ASK: Works in beforeEach(), but doesn't in beforeAll()
```
Yes, thanks for noticing! If I define relayConnectSpy in `beforeAll()` instead of `beforeEach()`, the mock just...doesn't work, the origin function is called, as if there has never been a mock defined.

```
beforeAll(() => {
  relayConnectSpy = jest.spyOn(DeveloperRelay.prototype, 'connect');
  relayConnectSpy.mockResolvedValue(new Duplex());
});
```

Hexxeh replied:
Can you point me to a test that fails in this scenario so I can help you debug it?

rafasofizada replied:
Almost every single one :) Because the mock simply doesn't work, every time `doConnect` is called (and it's called before/in every test), the `developerRelay.connect(host.id)` in `connectAction` sends an actual request to `https://api.fitbit.com:443/1/user/-/developer-relay/hosts/<hostType>`, which obv gives a 404

rafasofizada replied:
If you need, you can choose any except "logs an error if no ..." and "does not auto-connect a busy host"

rafasofizada replied:
`jest.config.js`: `restoreMocks: true` caused this.

Hexxeh replied:
Noticed you'd resolved this comment, but the `ASK` comment is still present. Remove it and then resolve this again.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -0,0 +1,56 @@
+  // Why spawn instead of fork? Fork() doesn't support unref()
+  const relayChildProcess = child_process.spawn('node', [relayJsPath], {
+    detached: true,
+    /**
```
Seems like you still have TODOs here to resolve? This comment reads a bit like Node docs, can likely be thinned down a bit once you're done... :)

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/26ee1d9e19334aedfd889137a6271b563b290a65

rafasofizada committed:
spawn() comments and minor test refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -0,0 +1,183 @@
+    const interval = 1000;
+    const readRelayInfo = jest.fn().mockReturnValue(undefined);
+
+    await pollRelayInfo(timeout, interval, readRelayInfo).catch(() => {});
```
You can use `expects().rejects` here.

rafasofizada replied:
What matcher can I use after `rejects` if all I am expecting is for the function to reject?

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/eee559877e708954b84b673a9b1f52bc66470989

Hexxeh replied:
`.rejects.toThrow()`

rafasofizada committed:
pollRelayInfo – replace p-wait-for, refactoring

p-wait-for uses ESM modules, which caused lots of problems when compiling the project. Switched to an alternative, async-wait-until.

readRelayInfo was injected as a dependency into pollRelayInfo, to enable easier testing. After refactoring of pollRelayInfo tests, no need to pass readRelayInfo as an argument to pollRelayInfo.

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Remove unnecessary tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo test minor refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
ESM-style fs/promises imports

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor relayInfo test refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
launch(), instance() refactoring & tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Check if local-relay dependency installed

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Better logging/UX

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented:
Looks like CI is failing still with a lint error this time. The commit hook should be notifying you of this when you commit: can you confirm it's firing when you commit?

Hexxeh commented on packages/sdk-cli/package.json:
```diff
@@ -22,6 +22,7 @@
     "@fitbit/portable-pixmap": "^1.0.3",
     "@moleculer/vorpal": "^1.11.5",
     "@openid/appauth": "^1.2.8",
+    "async-wait-until": "^2.0.8",
```
Looks like you've replaced `p-wait-for` but left all the Jest changes relevant to it in place, you should revert those if they're no longer necessary, particularly since they make reference to the unused/uninstalled package.

Hexxeh commented on packages/sdk-cli/src/models/DeveloperRelay.test.ts:
```diff
@@ -0,0 +1,186 @@
+    endpointMock = nock(apiUrl);
+  });
+
+  afterAll(() => {
```
Where are these mocked manually in order to need to be manually restored after all tests?

Hexxeh commented:
Created `feature/local-relay` as a feature branch to merge this to.

Hexxeh commented on packages/sdk-cli/src/models/HostConnections.test.ts:
```diff
@@ -23,22 +24,24 @@ function mockSentinel(spy: jest.SpyInstance) {
+      mockWS = mockSentinel(relayConnectSpy) as jest.Mocked<Duplex>;
       mockWS.destroy = jest.fn();
 
       mockRemoteHost = mockSentinel(remoteHostSpy);
```
I can't comment on the specific line because it hasn't changed, but the `acquires a developer relay connection for the given host ID` test doesn't really make sense anymore: it's testing test code. There's probably a new place that test belongs based on the refactoring you've done.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -0,0 +1,72 @@
+  });
+});
+
+describe('readJsonFile – actual file system', () => {
```
Why the FS mocking above if you're keeping the actual FS versions of these tests?

rafasofizada replied:
I...don't know. I didn't even mock it, it seems like I just wrapped the actual implementation around a `jest.fn()`. Removed

Hexxeh commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -0,0 +1,72 @@
+
+    await expect(readJsonFile(path)).rejects.toThrowError();
+
+    await fsPromises.unlink(path);
```
If the tests fails, this will never run and the file will be left on the filesystem. Things like this are why mock FS is a better idea as we discussed, it gives you a clean slate.

rafasofizada replied:
Hm, I thought we agreed that it's okay to use real-fs in these tests. How guaranteed is it that `afterEach/All` will run, even if the test fails or unexpectedly throws an error? I can either move the cleanup to `afterEach` or just replace the real-fs tests with mock-fs

Hexxeh replied:
I can't remember exactly what was said because it was a Meet call iirc, but I think the gist was either is okay, but both doesn't make sense, though mock FS gives a better clean slate for tests.

Hexxeh replied:
`after{Each,All}` will be called even if the case fails as far as I know, it's code within the test body that may not depending where it fails.

Hexxeh replied:
I would recommend mock FS.

rafasofizada committed:
Minor cleanup

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Test message fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
util.test.ts test refactor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Mock FS in util.test.ts

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Lint fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh committed:
Merge branch 'master' into local-relay

rafasofizada committed:
Remove jest config changes introduced in f57bdf1

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Make launch sync

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Playing with timers to prevent timeouts leaking out of jest tests

Jest warned:

"A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --runInBand --detectOpenHandles to find leaks."

because, apparently, timeout callbacks set in pollRelayInfo were executing after the tests were teared down.

This is a very hacky solution to the problem and _doesn't really work_ – will have to discuss.

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.test.ts:
```diff
@@ -42,6 +42,7 @@ describe('instance', () => {
       );
 
       await expect(instance()).resolves.toEqual(relayInfo);
+      // launch() is an empty moc
```
typo?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -72,6 +73,10 @@ describe('readRelayInfo', () => {
 
 describe('pollRelayInfo', () => {
+  beforeAll(() => {
+    // jest.useFakeTimers();
```
unused code?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -141,25 +148,30 @@ describe('pollRelayInfo', () => {
 
-    readJsonFileSpy.mockReset();
+    readJsonFileSpy.mockRestore();
+    jest.useRealTimers();
```
This will leak if any of the lines above fail and keep fake timers for other tests. Reset in an afterEach (it's a no-op if real timers are already being used iirc).

rafasofizada committed:
Typo

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
are pollRelayInfo tests causing CI errors?

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Kill subprocess in launch test

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Restore relayInfo tests

Still need refactoring (timeout mocks), but at least we know they weren't affecting the CI

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
launch test refactoring – less focus on node API, more focus on logic, use real files

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
instance – error fixes, better process/error handling/listening to events, extensive testing

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -72,108 +72,100 @@ describe('readRelayInfo', () => {
+   * The request is successful, so the interval should be cleared and value returned.
+   * Therefore, we check whether the interval is blocked whenever there is an existing request in-flight.
+   */
+  it('has only 1 in-flight poll request at a time', async () => {
```
I think we had this comment previously, but this test doesn't prove this, only the total number of calls.

Also use fake timers so this test doesn't take 1 second by itself. CI as a whole takes approx 30s even on the slow GitHub actions machines so this test is ~3% of the whole runtime alone out of hundreds.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -72,108 +72,100 @@ describe('readRelayInfo', () => {
+        new Promise<false>((resolve) => setTimeout(() => resolve(false), 2000)),
+    );
+
+    const poll = pollRelayInfo(timeout, undefined);
```
Also here.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -72,108 +72,100 @@ describe('readRelayInfo', () => {
+   * setInterval/Timeout/etc. are not guaranteed to call functions at __exactly__ the time specified,
+   * so there is a margin of 100ms (2 x 200ms + 100ms).
+   */
+  it('polls in regular intervals', async () => {
```
This also isn't testing what it claims. Again, mock timers are the only way you can test this.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.ts:
```diff
@@ -1,17 +1,24 @@
 import { ChildProcess, spawn } from 'child_process';
 import { Writable } from 'stream';
 
+// TODO: Separate log and error files?
```
Remove this before merge.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -1,54 +1,101 @@
+
+  it('spawns a process and logs output to a log file', (done) => {
+    const logOutput = 'test output';
+    logFilePath = join(__dirname, './launch-test-output.txt');
```
Put this in tmpdir just to be safe?

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/ce0aefa1c87fa79ad086c7f0ae9a974469a464bb
https://github.com/Fitbit/developer-bridge/pull/176/commits/026948b525cd4b37caf3754098ad228d0f7b284c
https://github.com/Fitbit/developer-bridge/pull/176/commits/5aa37c699115d9fa3b28ffad8930575bac36741e

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -1,54 +1,101 @@
+      const nodeArgs = ['-e', `console.log('${logOutput}')`];
+      subprocess = launch(nodeArgs, fd);
+
+      subprocess.on('error', (error) => {
```
nit: can you just do `subprocess.on('error', done);`?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -1,54 +1,101 @@
-        "child_process.spawn()'ed subprocess doesn't have stderr",
-      );
-    }
+      subprocess.on('error', (error) => {
```
This test is almost identical to the one above, factor it out into a function?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -29,9 +29,28 @@ export async function instance(): Promise<RelayInfo> {
+      const relayProcess = launch([relayJsPath], fd);
+
+      relayProcess.on('error', (error) => {
+        console.error('Local relay process threw error');
```
Include the error in the log?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -29,9 +29,28 @@ export async function instance(): Promise<RelayInfo> {
+      console.error(
+        `Error creating an output stream to file at path: ${RELAY_LOG_FILE_PATH}`,
+      );
+      throw error;
```
What does this actually do, re-throwing? You're already in an event handler, is this not just going to create an unhandled error? Or is that the point?

Hexxeh replied:
Perhaps turn `createWriteStream(RELAY_LOG_FILE_PATH)` into a Promise:

```
const fd = await new Promise((resolve, reject) => {
  createWriteStream(RELAY_LOG_FILE_PATH)
    .on('open', resolve)
    .on('error', reject);
});
```

rafasofizada replied:
Tried to do something similar with the `subprocess` too, but too much hassle. Left it as is

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/c8b70fddcb1921f878d20f7e25c27e8d6c141005

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.test.ts:
```diff
@@ -7,6 +11,28 @@ jest.mock('fs');
+  const addListener = stream.on.bind(stream);
+
+  jest.spyOn(stream, 'on').mockImplementation((listenerEvent, callback) => {
+    addListener(listenerEvent, callback);
```
Why not call `stream.on` here directly? Not clear why the bind is necessary.

rafasofizada committed:
Use stream instead of fd for subprocess logging

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Launch test refactor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
pollRelayInfo test refactor – use mock timers

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -3,6 +3,7 @@ import { ChildProcess } from 'child_process';
 import { join } from 'path';
 
 import { launch } from './launch';
+import { console } from 'fp-ts';
```
This seems like an accidental auto-import? I think you want console, the Node global?

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -83,87 +100,79 @@ describe('pollRelayInfo', () => {
+    // Checking if the behaviour is correct a couple of times is sufficient
+    const reps = 2;
+    // We need to ensure the timeout & interval actually allow for "a couple" of runs without timing out
+    if (timeout / interval < reps) {
```
This seems like the code of check that should be in the implementation, not the test, with the test checking for the throw?

Hexxeh replied:
Perhaps flip it and calculate interval from repls so it's guaranteed to be correct rather than checking the test itself?

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/5361e00a6a230680cf9b361f1b418b5ebe087d37

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -28,74 +29,44 @@ describe('launch', () => {
+        subprocess = launch(nodeArgs, fd);
+
+        subprocess.on('error', (error) => {
+          return done!(error);
```
Does this ever get fired? Both cases look happy path so I don't think so?

Hexxeh replied:
I'm not saying that to remove it, I can see you're trying to catch and fail the test if something goes wrong. Just hard to know it works as intended without it actually running.

Hexxeh replied:
You would have to have a bad executable path or similar to cause this.

rafasofizada replied:
Nope, I've tried it before and all the following things didn't work (i.e. they either output the error to subprocess.stderr, or didn't have an effect at all):

https://nodejs.org/api/child_process.html#event-error

For option 1:
- Pass bad executable path: `node randomrandom`
- Pass bad option/parameter: `node --randomradnom`

For option 2:
- `subprocess.kill('INVALID_SIGNAL')`

For option 3:
- `subprocess.send()` after subprocess closed or finished

Hexxeh replied:
Interesting, leave as is then.

rafasofizada replied:
Searching for "emit('error'" on https://github.com/nodejs/node/blob/master/lib/internal/child_process.js, it seems like `'error'` is reserved for only very corner cases, majority of other erroneous cases are handled directly, without emitting 'error'. I.e. it happens if you set a IPC channel on your spawned process, then disconnect the channel and send a message – will emit an error telling you that you can't deliver messages over a closed IPC channel.

rafasofizada committed:
Comments

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Output test logs to relay's tmpdir, + subtle bug fix

If the spawned node script threw an error (which we wouldn't expect), the error message could be something like:

```
console.lg(<EXPECTED OUTPUT>);
        ^

TypeError: console.lg is not a function
    at [eval]:1:9
    at ...
```

Because the error message quotes verbatim the line that contains <EXPECTED OUTPUT>, our test would pass, even if the script inside failed somehow. Therefore, exact match instead of regex match

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Ensure local-relay's output directory exists + DeveloperRelay.create() test

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo "regular interval" test refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
launch() returns Promise resolving on 'spawn' event

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
instance() refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
DeveloperRelay mkdir error handling + testing

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo timeout test – Jest timeout fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/index.ts:
```diff
@@ -29,32 +41,19 @@ export async function instance(): Promise<RelayInfo> {
-    relayProcess.on('close', async () => {
-      console.warn('Local relay process exited and closed');
-    });
+    throw error;
```
Again, I'm not sure throwing inside an error handler actually does anything? It might re-raise the error handler perhaps, but given you're already in it?

rafasofizada replied:
I've tested this and it seems to be like it behaves just like any other error throw, i.e. it will just abruptly end the process. I'm not very sure what the process is exactly though...is it only the CLI shell that is spawned by `npx fitbit` or is it `npx fitbit`, too?

Hexxeh replied:
It'll maybe end the process because Node sees an uncaught error, but that's not a particularly clean way of exiting. `npx fitbit` _is_ the CLI shell (sdk-cli Vorpal shell)

rafasofizada replied:
Log error and `process.exit(negative_number)` then?

Hexxeh replied:
Move this code into `launch.ts` and add the rejection case suggested there, at which point it'll be thown via the await at line 45.

rafasofizada replied:
`'error'` can be emitted at any point in future though; if I move it to `launch` and add the event listener inside the Promise, the Promise will reject **only** if `'error'` is emitted on *start*, which is only 1 of the possible `'error'` emission cases: https://nodejs.org/api/child_process.html#event-error

Hexxeh replied:
But do the other two cases apply? For 2. I think you're only killing in this error case and 3. you aren't using pipes to communicate.

rafasofizada replied:
Hmm right. That certainly makes things easier, thanks

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/680048b82c1fae3414b669d2fd23a4ced3669690

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.ts:
```diff
@@ -25,5 +24,7 @@ export function launch(
   relayChildProcess.unref();
 
-  return relayChildProcess;
+  return new Promise<ChildProcess>((resolve) =>
```
To be safe, add a rejection on `error`.

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/680048b82c1fae3414b669d2fd23a4ced3669690

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/680048b82c1fae3414b669d2fd23a4ced3669690

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.test.ts:
```diff
@@ -161,19 +154,26 @@ describe('pollRelayInfo', () => {
+          }),
+        );
+
+        return done();
```
nit: I think you could use `.rejects.` on the result of the promise here to make this a little clearer

rafasofizada replied:
That was actually the case before this commit. The issue is, I think using promises like this:

```
const promise = fnThatUsesTimeoutsAndPromises();
jest.runAllTimers();
await expect(promise).rejects...
```

is what causes the

```
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: Timed out after waiting for 240 ms".]
```

It seems like, indeed, the error in the prev CI run was caused by this test, now it is fixed, and the same error is now causes by another test, 'polls in regular intervals'. I will try changing the syntax to the current `then-catch` one, we'll see if the CI succeeds; if it doesn't, I'll change the syntax to `await expect().rejects` and will need to continue looking into what causes the error.

rafasofizada replied:
Are you sure about this? Because this doesn't seem like a Jest timeout error – this error message is from `async-wait-for` library, and it seems like the promise was rejected without a `try-catch` or `.catch()` clause

Hexxeh replied:
This error?

```[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: Timed out after waiting for 240 ms".] {
  code: 'ERR_UNHANDLED_REJECTION'
}```?

It comes from Node, not Jest.

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/1b5288515886cabe2144f191fece5e902354df49

Hexxeh commented on packages/sdk-cli/src/models/DeveloperRelay.ts:
```diff
@@ -41,7 +40,14 @@ export default class DeveloperRelay {
     if (local) {
       // Create an output directory (if doesn't exist)
-      await fsPromises.mkdir(RELAY_DIRECTORY_PATH);
+      try {
```
Creating the directory in the calling code for making a relay instance seems a little odd from an encapsulation perspective. I'd really like to see this PR land soon though, so if this is likely to be non-trivial for you to adjust, please just create a ticket to come back and refactor it.

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/176/commits/134923f23e7f0c7945d8ec7e430076a4c5208556`
`https://github.com/Fitbit/developer-bridge/pull/176/commits/1b3add13161c43a7f2fd4d1b125df4c5563972fe`

gmaniak commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -0,0 +1,42 @@
+    it('number string', () => expect(isInt('5')).toBe(false));
+  });
+
+  describe('true', () => {
```
NIT: Maybe to be more specific add a few more cases here. Examples that spring to mind are:
* Float
* Negative number

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/ce942bc655511107cf04d27c00c93403e9ae3c83

gmaniak commented on packages/sdk-cli/src/models/localRelay/relayInfo.ts:
```diff
@@ -0,0 +1,93 @@
+  // so this is standard way to check for installed dependencies.
+  try {
+    const { dependencies, devDependencies } = (await readJsonFile(
+      join(process.cwd(), 'package.json'),
```
NIT: use define for `package.json`

rafasofizada replied:
What is that?

Hexxeh replied:
Just a constant.

gmaniak replied:
Yes, just a constant, sorry for the confusion, C habits 🙈

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/1b5288515886cabe2144f191fece5e902354df49

rafasofizada replied:
`https://github.com/Fitbit/developer-bridge/pull/176/commits/e5a905b2808c14ccf33cb65dcf30872a3715e108`

gmaniak replied:
I was thinking of adding a constant for the `package.json` file name as it is used in multiple places, the path part was ok. Example:
```
const packageConfig = 'package.json' // or any other name you find appropriate
...
join(process.cwd(), packageConfig),
...
```

rafasofizada replied:
Oh! `package.json` is the standard (and I think required?) config file for all Node.js projects, so it's not something that is expected to be changed/configurable in the future, and it's always at the root of projects

rafasofizada replied:
I'll leave this unresolved and add a TODO to fix it later; rn have too much stuff going on and tbh I'm scared of touching any parts of the repo

gmaniak commented on packages/sdk-cli/src/models/DeveloperRelay.ts:
```diff
@@ -0,0 +1,107 @@
+        await fsPromises.mkdir(localRelay.RELAY_DIRECTORY_PATH);
+      } catch (error) {
+        if ((error as NodeJS.ErrnoException).code !== 'EEXIST') {
+          throw new Error('Error creating a directory for Local Relay files');
```
Is it actually an error if the directory already exists?

rafasofizada replied:
Just checked it through node repl, it is indeed. I think with `option.recursive` it won't throw an error but afaik it has some edge cases which will be too much effort to deal with

```
 [Error: EEXIST: file already exists, mkdir '/Users'] {
  errno: -17,
  code: 'EEXIST',
  syscall: 'mkdir',
  path: '/Users'
}
```

gmaniak replied:
Are we doing any directory cleanup somewhere or the directory needs to be deleted manually after each run?

rafasofizada replied:
Sorry, accidentally resolved. The directory is supposed to contain logs and a PID file, the PID file _should_ be cleaned up automatically by the local-relay process (if the process fails to do so, say in case of a crash, the PID file in the dir needs to be deleted manually), but the logs should remain. So no, the directory doesn't need cleanup or manual deletion

gmaniak replied:
Resolving this as I see that this part of the code was removed.

Hexxeh commented on packages/sdk-cli/src/models/localRelay/relayInfo.ts:
```diff
@@ -0,0 +1,93 @@
+
+export type RelayInfo = { port: number; pid: number };
+
+export type ReadRelayInfoResult = RelayInfo | false;
```
In dynamically typed languages like JS (which TypeScript is laying static types on top of) this is much less common. It happens in languages like C mostly because the two memory layouts for a type like `RelayInfo | false` are totally different but since JS handles it for you it's fine (and makes errors less likely, since you can't accidentally use invalid `pid` values for example in this case.

rafasofizada replied:
Oh I've never thought of that. I think, as Liam said, the most common style in JS is:

```
const result = fn();
if (!result) { /* handle error */ }
```

gmaniak replied:
Agreed, the original approach seems more JS oriented.

rafasofizada committed:
Move child process' lifecycle event handling to launch()

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
relayInfo test UnhandledPromiseRejection fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -1,18 +1,30 @@
 
+import { mockStreamWithEventEmit } from './index.test';
+
+// Is there a more idiomatic way to do this? Just using .spy() without mocking
```
`jest.spyOn(child_process, 'spawn').mockImplementation()` doesn't work?

rafasofizada replied:
I get a "Cannot set property spawn of [object Object] which has only a getter" error, which is something babel-related I think. I followed this https://stackoverflow.com/a/53209819/6539857

Hexxeh replied:
Having read that, I think what you have here is fine. I can't say I'd ran into that before, it's a new one on me too!

rafasofizada committed:
isInt -> isPositiveInt

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
createWriteStream util fn – ensure path exists (create if doesn't); DeveloperRelay refactoring

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
launch test dir cleanup + error handling bug fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
DeveloperRelay refactoring (see prev. commit)

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/util.test.ts:
```diff
@@ -1,17 +1,39 @@
+      expect(isPositiveInt(test)).toBe(false);
+    });
+
+    it('number string', () => expect(isPositiveInt('5')).toBe(false));
```
Why is this repeated 3x?

rafasofizada replied:
Accident, removed

rafasofizada committed:
launch test fix – wait for subprocess to close to read log files

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -66,23 +66,24 @@ describe('launch', () => {
+        ).resolves.toBe(logOutput + '\n');
 
-        return done!();
+        return done?.();
```
I'd suggest you use a promise to avoid having to get Jest to give you the right type here... :)

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/bf60349e40c30c3fedf46bba41f5cb4018cd3e5e

rafasofizada committed:
minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -66,23 +66,24 @@ describe('launch', () => {
-      subprocess.on('error', (error) => {
-        return done!(error);
-      });
+      // What happened to https://github.com/nodejs/node/issues/1751?
```
TODO?

rafasofizada replied:
More like a note...ChildProcess class should be exposed as per the github issue, but it isn't :/ so I have to resort to constructor.name. Remove?

Hexxeh replied:
Add a TODO here to revisit, I think it might be old typings.

rafasofizada replied:
https://github.com/Fitbit/developer-bridge/pull/176/commits/8a45aea402ac49f40b7747dd1dda4f90dd71df16

Hexxeh commented on packages/sdk-cli/src/models/localRelay/launch.test.ts:
```diff
@@ -66,23 +66,24 @@ describe('launch', () => {
-        return done!(error);
-      });
+      // What happened to https://github.com/nodejs/node/issues/1751?
+      await expect(subprocessPromise).resolves.toHaveProperty(
```
Are you just trying to check that it's an instance of ChildProcess? If so, use `expects().resolves.toBeInstanceOf(ChildProcess)`?

rafasofizada replied:
The class is not exposed for me ¯\_(ツ)_/¯ It's an interface in @types/node

rafasofizada committed:
Promisify launch test subprocess close

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
todo

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
process.exit() on launch tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Skip launch tests

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Minor

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Platform-agnostic path checks in util test

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Platform-agnostic path checks in util test – fix

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
package.json const TODO

Signed-off-by: Rafael Sofi-zada <rafa.sofizadeh@gmail.com>

rafasofizada committed:
Github Actions

Signed-off-by: rafasofizada <rafa.sofizadeh@gmail.com>
